pkg/server/ring.go

```diff
@@ -0,0 +1,275 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/grafana/dskit/dns"
+	"github.com/grafana/dskit/flagext"
+	"github.com/grafana/dskit/grpcclient"
+	"github.com/grafana/dskit/kv"
+	"github.com/grafana/dskit/kv/codec"
+	"github.com/grafana/dskit/kv/memberlist"
+	"github.com/grafana/dskit/netutil"
+	"github.com/grafana/dskit/ring"
+	ringclient "github.com/grafana/dskit/ring/client"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health/grpc_health_v1"
+)
+
+const ringKey = "ring"
+const ringName = "unified_storage"
+const numTokens = 128
+const heartbeatTimeout = time.Minute
+
+var metricsPrefix = ringName + "_"
+
+func (ms *ModuleServer) initRing() (services.Service, error) {
+	if !ms.cfg.EnableSharding {
+		return nil, nil
+	}
+
+	if ms.cfg.MemberlistJoinMember == "" {
```

Comment:
nit: "Seed" nodes may not necessarily join existing memberlist cluster on their own, but wait for others to join them instead. Checking for non-empty join member can be too restrictive, but we can keep it as is and modify once we learn how to setup the memberlist and ring in production.